### PR TITLE
Spike an IConnectionFactory for the sockets transport

### PR DESCRIPTION
**Pull request by davidfowl**
- Made the unix domain sockets test use it

I'd like to have more than one implementation of the `IConnectionFactory` to make sure the API is sound. This is one potential implementation.

### CODE REVIEW

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        /// </returns>
+        public async ValueTask<ConnectionContext> ConnectAsync(EndPoint endPoint, CancellationToken cancellationToken = default)
+        {
+            var protocolType = endPoint is UnixDomainSocketEndPoint ? ProtocolType.Unspecified : ProtocolType.Tcp;
+
+            var socket = new Socket(endPoint.AddressFamily, SocketType.Stream, protocolType);
+
+            await socket.ConnectAsync(endPoint);
+
+            var connection = new SocketConnection(socket, memoryPool: null, PipeScheduler.ThreadPool, new SocketsTrace(_logger));
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs]: The pool and scheduler might be better as constructor arguments to this factory.

Comment (halter73) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs]: I say you get ahead of it and create and options object (though not necessarily ioptions) before we add too many params to the constructor.

File: src/Shared/Pipelines/PipeReaderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    public static class PipeReaderExtensions
```
Comment (halter73) [inline src/Shared/Pipelines/PipeReaderExtensions.cs]: Who's calling these extension methods?

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs (hunk 1 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs]: The test

Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs]: Yep

File: src/Shared/Pipelines/PipeReaderExtensions.cs (hunk 2 above)
Comment (davidfowl) [reply src/Shared/Pipelines/PipeReaderExtensions.cs]: Tests

**Comment by davidfowl** [comment]
@halter73 give this another look.

File: src/Servers/Kestrel/Transport.Sockets/src/SharedSocketOptions.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        /// </summary>
+        /// <remarks>
+        /// Defaults to true.
+        /// </remarks>
+        public bool NoDelay { get; set; } = true;
+
+        /// <summary>
+        /// The maximum size before the transport will stop proactively reading from the transport.
+        /// </summary>
+        public long? MaxReadBufferSize { get; set; } = 1024 * 1024;
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SharedSocketOptions.cs]: While this is a break from KestrelServerLimits, I think we should standardize on having null be the default value for nullable ints and longs so we can more reasonably change the defaults later. We can use long.MaxValue or -1 to represent unlimited if we feel that's necessary. This should be more consistent with PipeOptions, StreamPipeReaderOptions, and StreamPipeWriterOptions.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        private readonly ILogger _logger;
+        private readonly SocketClientOptions _options;
+
+        /// <summary>
+        /// Creates the <see cref="SocketConnectionFactory"/>.
+        /// </summary>
+        /// <param name="options">The options for this transport</param>
+        /// <param name="loggerFactory">The logger factory</param>
+        public SocketConnectionFactory(IOptions<SocketClientOptions> options, ILoggerFactory loggerFactory)
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs]: Do we want a client API to take IOptions? Do we expect this normally to be retrieved from a DI container?

File: src/Servers/Kestrel/Transport.Sockets/src/SharedSocketOptions.cs (hunk 3 above)
Comment (halter73) [reply src/Servers/Kestrel/Transport.Sockets/src/SharedSocketOptions.cs]: I forgot SocketTransportOptions still derives from this... I guess that means we shouldn't be changing the interpretation of null. It is an inconsistency with PipeOptions, but at leas this is more consistent with other Kestrel-related options I guess 😞

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs (hunk 4 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs]: Tough question but a good one. The only client API we have now is DI enabled.

**Comment by davidfowl** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=277157

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2293

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs (hunk 4 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionFactory.cs]: You know how we solve this? We leave it DI friendly and then add static factory methods for the non DI case.

`var factory = SocketConnectionFactory.Create();`

**Comment by davidfowl** [comment]
@anurse  I want to get something like this in 3.1. Lets discuss next week on possibilities.

**Pull request closed without merging**